crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3494,14 +3494,19 @@ impl<'db> TypeInferenceBuilder<'db> {
                 Type::Todo
             }
 
-            // TODO PEP-604 unions
             ast::Expr::BinOp(binary) => {
-                self.infer_binary_expression(binary);
```

Comment:
I struggled with this a bit -- should this line still be here? If it's left as-is, we hit the `assert!(previous.is_none());` assertion below. If we leave it as-is but use `let left_ty = self.expression_ty(&binary.left);` instead of `let left_ty = self.infer_type_expression(&binary.left);`, then we get inferences like `Literal[int] | Literal[None]`.

Reply:
I think the way you've done it in this PR is correct!

The invariant is that we should only insert a type into `self.types.expressions` _once_ for each AST node in the tree. What `self.infer_binary_expression()` does here is it iterates through all the subnodes and stores a type in `self.types.expressions` for each subnode. But that's incorrect with the changes you're making here, because we're calling `self.infer_type_expression` on each subnode, and `self.infer_type_expression` already stores a type in `self.types.expressions` for each subnode (it's the `let previous = self.types.expressions.insert(expr_id, ty);` line at the bottom of this method).